### PR TITLE
Build Scripted REST API to query G-Invoicing and Confirm Order Validity

### PR DESCRIPTION
**Pull request by zachclark-ccpo**
# Overview
This ticket fulfills the requirements for [AT-7627](https://ccpo.atlassian.net/browse/AT-7626), to query the G-Invoicing API to confirm if an Order Number exists in their records.
## Resources created
Several resources are created on SNOW with this PR:
- Scripted REST API **G-Invoicing** with a GET Request resource `/order_validation` which the frontend can use to determine the validity of an Order Number
image
- Server side Script Includes **gInvoicingRequest** which makes the request to G-Invoicing and returns the `http_payload`
image
# Scripts created
## `Order Validation`
```
(function process(/*RESTAPIRequest*/ request, /*RESTAPIResponse*/ response) {
	
	// Gather configuration values from ATAT Configuration Table
	var grAtatConfig = new GlideRecord("x_g_dis_atat_configuration");
	grAtatConfig.get("name", "ginvoicing_endpoint");
	var gInvoicingEndpoint = grAtatConfig.value.toString();
	grAtatConfig.get("name", "ginvoicing_atat_systemid");
	var atatSystemId = grAtatConfig.value.toString();
	var queryParams = request.queryParams;
	
	// Payload to send to gInvoicingRequest
	var payload = {
		atat_system_id: atatSystemId,
		endpoint: gInvoicingEndpoint,
		order_number: queryParams.orderNumber.toString()
		
	};
	var gInvoicingOrder = gInvoicingRequest(payload);
	if (gInvoicingOrder.valid){
		return {
			valid: true,
			message: "Order number exists in the G-Invoicing System of Record"
		};
	}
	else{
		response.setStatus(404);
		return {
			valid: false,
			message: "Order number doesn't exist in the G-Invoicing System of Record"
		};
	}
})(request, response);
```
## `gInvoicingRequest`
```
function gInvoicingRequest(payload){
	var http_payload;
	try{
		// Outbound REST message
		var request = new sn_ws.RESTMessageV2();
		
		// Request set-up
		var gInvoicingEndpoint = payload.endpoint + "/" + payload.order_number;
		request.setEndpoint(gInvoicingEndpoint);
		request.setHttpMethod("get");
		request.setLogLevel("all");
		request.setMutualAuth("g-invoicing-mauth");
		request.setRequestHeader("SystemID", payload.atat_system_id);
		
		// Make request
		var response = request.execute();
		// Check for errors
		if(response.getStatusCode() != "200"){
			gs.error(response.getStatusCode());
			gs.error(response.getBody());
			http_payload = {
				body: response.getBody(),
				status_code: response.getStatusCode(),
				headers: response.getAllHeaders(),
				valid: false,
			};
			return http_payload;
		}
		http_payload = {
			body: response.getBody(),
			status_code: response.getStatusCode(),
			headers: response.getAllHeaders(),
			valid: true,
		};
		return http_payload;
	}
	catch(error){
		gs.error("Internal SNOW error => " + error);
		gs.error("Error Stack => " + error.stack);
	}
}
```
# Testing
## Protocol and Certificate Set up
Before you can test the API, you will need to create a new Protocol Profile on SNOW:
image
This Protocol Profile requires a new Certificate, use the Cert @jeffsegal-ccpo sent on 7/13/2022. Configure this cert as a PKCS12 Keystore:
image
## Testing the API and function
Use **Explore REST API** to test the functionality
image
To test a valid response, add the following queryParameter to the request:
```
orderNumber: O2206-097-097-184790
```
image

### Expected Success Response
```
{
  "result": {
    "valid": true,
    "message": "Order number exists in the G-Invoicing System of Record"
  }
}
```
image

### Expected Failure Response
```
{
  "result": {
    "valid": false,
    "message": "Order number doesn't exist in the G-Invoicing System of Record"
  }
}
```
image

### CODE REVIEW

**Review by juliusfitzhugh-ccpo — APPROVED** [review]
Nice work putting together this G-Invoicing resource. This helps lay the ground work for other requests.

File: f600233d1b154d507b782f84604bcb12/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Generated files
```
Comment (jasonburkert-ccpo) [inline f600233d1b154d507b782f84604bcb12/README.md:1]: This file was moved by #102

**Review by jasonburkert-ccpo — DISMISSED** [review]
This looks good.  On the /order resource issue, I could go either way.  Thinking this is fine and we can adjust later if we have to look up GT&Cs

**Comment by jasonburkert-ccpo** [comment]
Got the successful response on PDI

image

File: f600233d1b154d507b782f84604bcb12/README.md (hunk 1 above)
Comment (zachclark-ccpo) [reply f600233d1b154d507b782f84604bcb12/README.md]: Will delete and re-request review

Comment (zachclark-ccpo) [reply f600233d1b154d507b782f84604bcb12/README.md]: Going to remove that file now.

**Pull request merged**